### ISSUE TITLE
Support for Razer Naga Epic

### ISSUE THREAD

**Issue by TheOneTheyCall-Tim** [stickoking/razer-macos#3]
Hey, first issue? Great! Per request from stickoking, I am copying my issue here: 

I have a Razer Naga Epic, and like a lot of you, am furious with Razer for abandoning Mac users. This is a gaming mouse, but I got a huge productivity boost in my CAD work thanks to the thumb keys. 

I have had good success with Karabiner Elements in programming the thumb keys. However, Karabiner does not do anything for adjusting the mouse dpi/sensitivity settings. What used to be a super sensitive mouse that I had to slow down to use effectively is now too slow. I have to pick it up and reposition it all the time, even though I have the macOS sensitivity maxed out.

I love what you are doing here. I was pleased to see a bunch of Nagas supported. However, the Epic was not on the list. I'm thinking mine must be from 2011 or 2012, but still works absolutely great. Sure enough, after installation, my Naga Epic was not recognized. I don't care so much about the lighting, but I was really hoping you could solve my inability to adjust the dpi sensitivity settings.

In further research looking at OpenRazer, I don't know enough to really know what I am looking at, but I see product IDs listed there for various products. Mine is not.

Here is the info for the mouse I have:

Razer Naga Epic:
Product ID: 0x001f
Vendor ID: 0x1532
Version: 2.00
Speed: Up to 12 Mb/s
Manufacturer: Razer
Location ID: 0x14113000 / 10
Current Available (mA): 500
Current Required (mA): 100
Extra Operating Current (mA): 0

**Comment by stickoking** [comment]
Will add support created a branch to work on this https://github.com/stickoking/razer-macos/tree/dev/razer_naga_epic

Will update here once it's done, I will first create a pre-release for you to test it with your device, as I do not own a Naga Epic to test it out. If all is good will move it to Release on 0.4.13

**Comment by TheOneTheyCall-Tim** [comment]
That would be fantastic! 
Happy to help in any way I can, just know that all my programing experience is from 30 years ago. You'll have to walk me through anything you need me to do.

**Comment by stickoking** [comment]
Hello Hello @TheOneTheyCall-Tim 

The Naga Epic is a wireless mouse yes? if its wireless can you provide the information you provided with product id one when in wired mode and one in wireless mode.

**Comment by TheOneTheyCall-Tim** [comment]
Yes, it is either wireless or wired. I have been using it wired for quite a while, as the battery went bad and were hard to come by for a long time. The above information is the wired set. 
I see there are some batteries running around now. I have been meaning to fix that for a long time anyway. I'll see if I can't get my hands on a new one by next weekend. 

I don't actually remember for sure, but I think wireless mode is a proprietary wireless signal between the mouse and it's base and not bluetooth. If that matters, I can try to look that up. I still have original documentation for it. There's a pairing button both on the mouse and on the base, so I expect that is the case.

**Comment by stickoking** [comment]
> Yes, it is either wireless or wired. I have been using it wired for quite a while, as the battery went bad and were hard to come by for a long time. The above information is the wired set. 
> 
> I see there are some batteries running around now. I have been meaning to fix that for a long time anyway. I'll see if I can't get my hands on a new one by next weekend. 
> 
> 
> 
> I don't actually remember for sure, but I think wireless mode is a proprietary wireless signal between the mouse and it's base and not bluetooth. If that matters, I can try to look that up. I still have original documentation for it. There's a pairing button both on the mouse and on the base, so I expect that is the case. 

Yup that's hence why I asked for the two different modes according to open Razer the wired and wireless mode have different ids so I need to confirm them I believe it's cause the dock reports as a the mouse as wireless mode but a different id

**Comment by TheOneTheyCall-Tim** [comment]
I wonder if I can get the information you need then just by plugging in the base, regardless of the mouse not having a battery?

**Comment by TheOneTheyCall-Tim** [comment]
Looks like that might work.
What was listed before as Razer Naga Epic is now listed as Razer Naga Epic Dock

Razer Naga Epic Dock:
  Product ID:	0x0021
  Vendor ID:	0x1532
  Version:	2.00
  Speed:	Up to 12 Mb/s
  Manufacturer:	Razer
  Location ID:	0x14113000 / 12
  Current Available (mA):	500
  Current Required (mA):	100
  Extra Operating Current (mA):	0

**Comment by stickoking** [comment]
Hello Hello @TheOneTheyCall-Tim 

I made a pre-release its for the wired version only right now 

please help me test it out 
https://github.com/stickoking/razer-macos/releases/tag/v0.4.14a

**Comment by TheOneTheyCall-Tim** [comment]
Ok. That does appear to recognize the mouse. 
However, it does not seem to let me have any control over it.
I can open the Razer Naga Epic window for custom color, DPI and Polling Rate. however, changing any of the values does not alter the mouse. The color never changes. Neither the speed nor sensitivity of the mouse changes. 

In the main drop down menu offering color effects such as spectrum, cycle, wave, breathing, etc...  none of those alter the mouse either. 


It's hard to tell, but I do think just the existence of this being installed has made the mouse a certain (fixed) amount more sensitive. That was one of my main complaints before, so I do believe this is already helping in that regard. 

Thank you so much for doing this. I think we are getting close!

**Comment by TheOneTheyCall-Tim** [comment]
I just rebooted to see if that made any difference. No. It does not. 

In messing around with the State Manager, I see the current device's state does match what I've given it: 
Razer Naga Epic
MODE          "breathe"
ARGS            [0]
DPI                5600
POLLRATE    1000

However, the mouse's actual behavior does not match. 

Also as an aside, the only State defined is All Lights Off, and I do not see the Create that is supposed to be in a dropdown menu or any other way to create/edit/duplicate/otherwise achieve any state other that All Lights Off.

**Comment by TheOneTheyCall-Tim** [comment]
After using this for a couple of days, I am not convinced there was any difference in DPI, fixed or otherwise. It certainly is not adjustable. 

So it does recognize the mouse by name, and I can open the State Manager for the Razer Naga Epic, but I do not see that it has any influence over the mouse.

**Comment by stickoking** [comment]
@TheOneTheyCall-Tim  

Thanks for your feedback it is likely I have gotten the wrong transaction id hence you're not seeing your configuration reflect on the device I will scour openrazer see if I can get the correct transaction id for this device. In regards to profiles for state management. It works this way

You first need to set the configuration you want then in the state dropdown you just type in any profile name you wish to save the state as 

Screenshot below

![image](https://github.com/stickoking/razer-macos/assets/41362698/d6d9f7d2-cb2e-4d49-bbdd-e460198f6052)

Then you can apply this state to different 'states' in the Settings tab

Screenshot below

![image](https://github.com/stickoking/razer-macos/assets/41362698/44341f0d-5a2c-44b3-bbf8-050312602cc2)


I will try to resolve the issue that settings not reflecting on ur device and update the pre-release

**Comment by TheOneTheyCall-Tim** [comment]
@stickoking , Thank you for spending time on this!

**Comment by TheOneTheyCall-Tim** [comment]
@stickoking , Hi. Not trying to push or anything, but just wondering if you've had a chance to look at this any further?
